src/Security/Authentication/test/CertificateTests.cs

```diff
@@ -0,0 +1,631 @@
+                    Events = failedValidationEvents
+                },
+                Certificates.SelfSignedValidWithServerEku);
+
+            var response = await server.CreateClient().GetAsync("https://example.com/");
+            Assert.Equal(HttpStatusCode.Forbidden, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task DoingNothingInTheValidationEventReturnsOK()
+        {
+            var server = CreateServer(
+                new CertificateAuthenticationOptions
+                {
+                    AllowedCertificateTypes = CertificateTypes.SelfSigned,
+                    ValidateCertificateUse = false,
+                    Events = unprocessedValidationEvents
+                },
+                Certificates.SelfSignedValidWithServerEku);
+
+            var response = await server.CreateClient().GetAsync("https://example.com/");
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task VerifyNotSendingACertificateEndsUpInForbidden()
+        {
+            var server = CreateServer(
+                new CertificateAuthenticationOptions
+                {
+                    Events = sucessfulValidationEvents
+                });
+
+            var response = await server.CreateClient().GetAsync("https://example.com/");
+            Assert.Equal(HttpStatusCode.Forbidden, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task VerifyArrHeaderIsUsedIfCertIsNotPresent()
+        {
+            var server = CreateServer(
+                new CertificateAuthenticationOptions
+                {
+                    AllowedCertificateTypes = CertificateTypes.SelfSigned,
+                    Events = sucessfulValidationEvents
+                },
+                wireUpHeaderMiddleware : true);
+
+            var client = server.CreateClient();
+            client.DefaultRequestHeaders.Add("X-ARR-ClientCert", Convert.ToBase64String(Certificates.SelfSignedValidWithNoEku.RawData));
+            var response = await client.GetAsync("https://example.com/");
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task VerifyArrHeaderEncodedCertFailsOnBadEncoding()
+        {
+            var server = CreateServer(
+                new CertificateAuthenticationOptions
+                {
+                    Events = sucessfulValidationEvents
+                },
+                wireUpHeaderMiddleware: true);
+
+            var client = server.CreateClient();
+            client.DefaultRequestHeaders.Add("X-ARR-ClientCert", "OOPS" + Convert.ToBase64String(Certificates.SelfSignedValidWithNoEku.RawData));
+            var response = await client.GetAsync("https://example.com/");
+            Assert.Equal(HttpStatusCode.Forbidden, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task VerifySettingTheHeaderOnTheForwarderOptionsWorks()
```

Comment:
Move forwarder tests to HttpOverride.Tests.

Reply:
I created a modified copy of these tests in the http override but I kept the end to end tests here as well, since we do want to test cert auth + the forwarder actually authenticates / forbids correctly